File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cs
public class Solution {
    public int MaxProfit(int[] prices) {
        int maxProfit = 0, buyingPrice = prices[0];
        for(int i =0;i<prices.Length;i++){
            maxProfit = Math.Max(maxProfit , prices[i] - buyingPrice);

            buyingPrice = Math.Min(buyingPrice, prices[i]);
        }
        return maxProfit;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        //dp or might be a greedy algorithm:
        //TC: O(n) , SC: O(1)
        int n = prices.size(), mini = prices[0], maxProfit = 0;
        for(int i = 1;i<n;i++){
            int profit = prices[i] - mini;
            maxProfit = max(profit, maxProfit);
            mini = min(mini, prices[i]);
        }
        return maxProfit;
    }
};